edelphi-persistence/src/main/java/fi/metatavu/edelphi/dao/users/UserDAO.java

```diff
@@ -246,5 +250,90 @@ public User updateLocale(User user, String locale) {
+    CriteriaQuery<User> criteria = criteriaBuilder.createQuery(User.class);
+    Root<User> user = criteria.from(User.class);
+    Subquery<Long> subquery = criteria.subquery(Long.class);
+    Root<DelfoiUser> delfoiUser = subquery.from(DelfoiUser.class);
+    subquery.select(delfoiUser.get(DelfoiUser_.id))
+      .where(criteriaBuilder.and(
+        criteriaBuilder.equal(delfoiUser.get(DelfoiUser_.user), user),
+        criteriaBuilder.equal(delfoiUser.get(DelfoiUser_.role), excludedRole)
+      ));
+
+    Subquery<Long> panelUserSubquery = criteria.subquery(Long.class);
+    Root<PanelUser> panelUser = panelUserSubquery.from(PanelUser.class);
+    panelUserSubquery.select(panelUser.get(PanelUser_.id))
+      .where(criteriaBuilder.equal(panelUser.get(PanelUser_.user), user));
+
+    criteria.select(user)
+      .distinct(true)
+      .where(criteriaBuilder.and(
+        criteriaBuilder.not(criteriaBuilder.exists(panelUserSubquery)),
+        criteriaBuilder.lessThan(user.get(User_.lastLogin), before),
+        criteriaBuilder.isFalse(user.get(User_.archived)),
+        criteriaBuilder.not(criteriaBuilder.exists(subquery))
+      ));
+
+    return entityManager.createQuery(criteria).setMaxResults(maxResults).getResultList();
+  }
+
+  public List<User> listUsersToDelete(Date before, int maxResults) {
+    EntityManager entityManager = getEntityManager();
+    CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+    CriteriaQuery<User> criteria = criteriaBuilder.createQuery(User.class);
+    Root<User> root = criteria.from(User.class);
+    criteria.select(root);
+
+    criteria.where(
+      criteriaBuilder.and(
+        criteriaBuilder.lessThan(root.get(User_.lastModified), before),
+        criteriaBuilder.equal(root.get(User_.archived), Boolean.TRUE)
+      )
+    );
+
+    return entityManager.createQuery(criteria).setMaxResults(maxResults).getResultList();
+  }
+
+  public List<User> listAllByCreator(User user) {
+    EntityManager entityManager = getEntityManager();
+
+    CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+    CriteriaQuery<User> criteria = criteriaBuilder.createQuery(User.class);
+    Root<User> root = criteria.from(User.class);
+    criteria.select(root);
+    criteria.where(
+      criteriaBuilder.equal(root.get(User_.creator), user)
+    );
+
+    return entityManager.createQuery(criteria).getResultList();
+  }
+
+  public List<User> listAllByModifier(User user) {
+    EntityManager entityManager = getEntityManager();
+
+    CriteriaBuilder criteriaBuilder = entityManager.getCriteriaBuilder();
+    CriteriaQuery<User> criteria = criteriaBuilder.createQuery(User.class);
+    Root<User> root = criteria.from(User.class);
+    criteria.select(root);
+    criteria.where(
+      criteriaBuilder.equal(root.get(User_.lastModifier), user)
+    );
+
+    return entityManager.createQuery(criteria).getResultList();
+  }
+
+  @Override
```

Comment:
Overriding the findById is not a good idea. Check the archived flag on the controller level instead